### PR TITLE
Optimize incremental summation of unique symbols using CustomAdd sympy expression. 

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139867

**wins**
on torchrec benchmark, for 2K nodes we go from 333 to 295 (13% or so)
```
buck2 run @fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=200
```
This diff optimizes construction expressions of the form 
a+b+c...  (all unique symbols). 
which are very common in torchrec models. 

**How**
Expressions of the form a+b+c are not optimized by add, the only needed optimization is sorting them.
If we have  a+b+c and we are adding (d) to it, we can do a binary search to know 
the position of (d) and avoid optimizing the new expression by passing the new order. 

CustomAdd subclass sympy.Add and does some logic to assign attributes 
that store information about the expression and use it to build the next
expression when a symbol is added with out having to run all sympy.Add optimizations.

**Extensions**:
1. support constant terms.
2. support 10a+10b+.. (this will give even more wins will extend the support in second PR)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @SherlockNoMad @EikanWang @wenzhe-nrv

### CODE REVIEW

**Comment by laithsakka** [comment]
slow down is because i have run_extra_validations turned on not an issue

File: torch/utils/_sympy/functions.py [hunk 1]
```diff
@@ -841,6 +841,141 @@ def _find_localzeros(cls, values, **options):
...
+        rhs = args[1]
+
+        def make_optimized(ordered_args_as_list):
+            obj = super(CustomAdd, cls).__new__(
+                cls, *ordered_args_as_list, evaluate=False
+            )
+            assert isinstance(obj, CustomAdd)
+            obj.ordered_summation_of_unique_symbols = True
+
+            # if True : # config.run_extra_validations: # need to rebase on another PR to see this lol
```
Comment (laithsakka) [inline torch/utils/_sympy/functions.py:888]: will rebase and uncomment this once the https://github.com/pytorch/pytorch/pull/139741 lands

**Comment by laithsakka** [comment]
now this 2% sounds legit lol

**Comment by laithsakka** [comment]
ok so regarding the pr_time_benchmark regression it runs now with N=100 for that i can measure 1.5% regression 
but if we increase that to N=1000 I measure 40% win ! 
so two things:
1. I will try to make the overhead for small N less
2. we need larger N for the benchmark for sure. !

**Comment by laithsakka** [comment]
N=1000
with the change
```
compile time instruction count for iteration 1 is 61047258411
compile time instruction count for iteration 2 is 61054004626

```

with out 
```
compile time instruction count for iteration 1 is 105381896846
compile time instruction count for iteration 2 is 105380498639
```

**Comment by laithsakka** [comment]
The source of the 1-2% overhead is the check in here, will see how i can make it better.
converting CustomAdd to Add basically
    def expr(self):
        return self.expr_allow_custom_add
        # We make sure that CustomAdd is not leaking outside the construction loop.
        # This is needed to allow Add(a, b) == CustomAdd(a, b) to be true.
        if isinstance(result, torch.utils._sympy.functions.CustomAdd):
            import sympy

            return sympy.Add(*result._args, evaluate=False)

        return result

File: torch/fx/experimental/sym_node.py [hunk 2]
```diff
@@ -170,6 +170,18 @@ def _value_hash(self) -> int:
 
     @property
     def expr(self):
+        result = self.expr_allow_custom_add
+        # We make sure that CustomAdd is not leaking outside the construction loop.
+        # This is needed to allow Add(a, b) == CustomAdd(a, b) to be true.
+        if isinstance(result, torch.utils._sympy.functions.CustomAdd):
+            import sympy
+
+            return sympy.Add(*result._args, evaluate=False)
```
Comment (laithsakka) [inline torch/fx/experimental/sym_node.py:179]: this is the source of the 2% regression when N=100
when N=1000 we have 60% win though!

File: torch/fx/experimental/sym_node.py [hunk 3]
```diff
@@ -1134,6 +1152,8 @@ def binary_magic_impl(self, other):
                     out = Mod(self.expr, other.expr)
                 else:
                     out = PythonMod(self.expr, other.expr)
+            elif method == "add":
+                out = func(self.expr_allow_custom_add, other.expr_allow_custom_add)
```
Comment (laithsakka) [inline torch/fx/experimental/sym_node.py:1156]: this is wrong lol

**Comment by laithsakka** [comment]
I marked this unready because i want to investigate what does it take to fix those and make CustomAdd livable with Add
https://github.com/pytorch/pytorch/pull/139943

File: torch/fx/experimental/sym_node.py (hunk 2 above)
Comment (laithsakka) [reply torch/fx/experimental/sym_node.py:179]: actually this is not acceptable it reduce savings by 20s, looking at avoiding this.

**Comment by ezyang** [comment]
@isuruf what I told Laith is that this seemed like it was correct but it might be better to go straight to a new version of Add that maintains internal coefficient dictionary so that insertions are quick.

@laithsakka your rewrite is going to redo this completely, no? Because you wouldn't need to do binary search inserts anymore. So I don't think there is much utility reviewing this.

**Comment by isuruf** [comment]
> @isuruf what I told Laith is that this seemed like it was correct but it might be better to go straight to a new version of Add that maintains internal coefficient dictionary so that insertions are quick.

Right. I looked at adding this in sympy itself a month or so ago. Let me see how much improvement I get with pr_time_benchmark with that branch

**Comment by laithsakka** [comment]
> @isuruf what I told Laith is that this seemed like it was correct but it might be better to go straight to a new version of Add that maintains internal coefficient dictionary so that insertions are quick.
> 
> @laithsakka your rewrite is going to redo this completely, no? Because you wouldn't need to do binary search inserts anymore. So I don't think there is much utility reviewing this.

The rewrite will just move the attribute back to the SymNode, it will be similar but there is no need to review until I publish it

**Comment by laithsakka** [comment]
> > @isuruf what I told Laith is that this seemed like it was correct but it might be better to go straight to a new version of Add that maintains internal coefficient dictionary so that insertions are quick.
> 
> Right. I looked at adding this in sympy itself a month or so ago. Let me see how much improvement I get with pr_time_benchmark with that branch

if you have the PR I can test it on the Model that I am playing with. you can also try the sum_floordiv_regression with much more nodes (increase N)

**Comment by isuruf** [comment]
I just checked. The optimization where the internal dictionary is maintained was slower than sympy master. However I have some optimizations at https://github.com/isuruf/sympy/tree/micro that are good wins. Can you try that branch?

**Comment by laithsakka** [comment]
> I just checked. The optimization where the internal dictionary is maintained was slower than sympy master. However I have some optimizations at https://github.com/isuruf/sympy/tree/micro that are good wins. Can you try that branch?

i tested it on the sum_floor_div with 1000 Nodes, I do see 40%-50% wins on that benchmark with your chanages
that said
N=1000
96G instructions to -> 50G instructions. 


running this diff on top of those opts further reduce the time 
50G instructions. -> 44L instructions .

12% win, is it worth it? 
This is does not answer the actual effect on benchmark , the OSS benchmark broken, but maybe  
I can apply the changes to third party to evaluate?

@isuruf what would be your estimate time of when those opts will be available?
and do you know what specific changes I can apply manually ?

I think I would proceed with those for the 12% win for now. the other diff seems really simple most of it is unit test.
https://www.internalfb.com/diff/D65837755

**Comment by laithsakka** [comment]
actually better knews :) i ported the changes to sympy to third party and tested  with 2K nodes, 
and the changes of @isuruf have the same effect as this diff. 
the only remaining thing to things to check is if it handles the co-efficient extension well.(I will check that tomorrow). (12*c+12*b...)

There is a good probability you dont have to review the binary search ;) @ezyang

**Comment by laithsakka** [comment]
closing this for the favor of this
https://www.internalfb.com/diff/D65837755

**Pull request closed without merging**